File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/model/Organization.kt
package com.arcinteractive.spaces.data.model

import java.util.Date

data class Organization(
    val id: String,
    val name: String,
    val status: OrganizationStatus,
    val entitlements: OrganizationEntitlements,
    val usage: OrganizationUsage,
    val createdAt: Date?
)

enum class OrganizationStatus(val wireValue: String) {
    Active("active"),
    Suspended("suspended")
}

data class OrganizationMember(
    val id: String,
    val userId: String,
    val displayName: String,
    val email: String?,
    val role: OrganizationRole,
    val status: OrganizationMemberStatus,
    val joinedAt: Date?
)

enum class OrganizationRole(val wireValue: String) {
    PrimaryAdministrator("primary_admin"),
    Administrator("admin"),
    Member("member");

    val canManageOrganization: Boolean
        get() = this == PrimaryAdministrator || this == Administrator
}

enum class OrganizationMemberStatus(val wireValue: String) {
    Active("active"),
    Suspended("suspended")
}

data class OrganizationEntitlements(
    val peopleCapacity: Int?,
    val activeSpaceCapacity: Int?,
    val enabledModuleIds: Set<String>,
    val mediaStorageCapacityBytes: Long?
) {
    companion object {
        val Foundation = OrganizationEntitlements(
            peopleCapacity = 250,
            activeSpaceCapacity = 10,
            enabledModuleIds = setOf("general", "events", "polls", "members", "settings"),
            mediaStorageCapacityBytes = 10L * 1024L * 1024L * 1024L
        )

        fun effective(context: android.content.Context, stored: OrganizationEntitlements): OrganizationEntitlements {
            return stored
        }
    }
    fun allowsAddingPerson(currentPeople: Int): Boolean {
        val capacity = peopleCapacity ?: return false
        return capacity > 0 && currentPeople < capacity
    }

    fun allowsActivatingSpace(currentActiveSpaces: Int): Boolean {
        val capacity = activeSpaceCapacity ?: return false
        return capacity > 0 && currentActiveSpaces < capacity
    }

    fun allowsModule(moduleId: String): Boolean = enabledModuleIds.contains(moduleId)

    fun allowsStorageIncrease(currentBytes: Long, additionalBytes: Long): Boolean {
        val capacity = mediaStorageCapacityBytes ?: return false
        if (currentBytes < 0 || additionalBytes < 0) return false
        return currentBytes <= capacity - additionalBytes
    }
}

data class OrganizationUsage(
    val peopleCount: Int,
    val activeSpaceCount: Int,
    val mediaStorageBytes: Long
) {
    fun afterStorageRelease(bytes: Long): OrganizationUsage = copy(
        mediaStorageBytes = maxOf(0L, mediaStorageBytes - maxOf(0L, bytes))
    )
}

data class OrganizationOwnedSpace(
    val id: String,
    val name: String,
    val emoji: String,
    val memberCount: Int,
    val memberIds: List<String>,
    val isArchived: Boolean
)

data class OrganizationInvite(
    val id: String,
    val organizationId: String,
    val organizationName: String,
    val role: OrganizationRole
)

File: iOS/Spaces V1.6/Models/Organization.swift
import Foundation

struct Organization: Identifiable, Hashable {
    let id: String
    let name: String
    let status: OrganizationStatus
    let entitlements: OrganizationEntitlements
    let usage: OrganizationUsage
    let createdAt: Date?
}

enum OrganizationStatus: String, Hashable {
    case active
    case suspended
}

struct OrganizationMember: Identifiable, Hashable {
    let id: String
    let userID: String
    let displayName: String
    let email: String?
    let role: OrganizationRole
    let status: OrganizationMemberStatus
    let joinedAt: Date?
}

enum OrganizationRole: String, CaseIterable, Hashable {
    case primaryAdministrator = "primary_admin"
    case administrator = "admin"
    case member

    var canManageOrganization: Bool {
        self == .primaryAdministrator || self == .administrator
    }
}

enum OrganizationMemberStatus: String, Hashable {
    case active
    case suspended
}

struct OrganizationEntitlements: Hashable {
    let peopleCapacity: Int?
    let activeSpaceCapacity: Int?
    let enabledModuleIDs: Set<String>
    let mediaStorageCapacityBytes: Int64?

    static let foundation = OrganizationEntitlements(
        peopleCapacity: 250,
        activeSpaceCapacity: 10,
        enabledModuleIDs: ["general", "events", "polls", "members", "settings"],
        mediaStorageCapacityBytes: 10 * 1_024 * 1_024 * 1_024
    )

    static func effective(stored: OrganizationEntitlements, defaults: UserDefaults = .standard) -> OrganizationEntitlements {
        return stored
    }

    func allowsAddingPerson(currentPeople: Int) -> Bool {
        guard let peopleCapacity, peopleCapacity > 0 else { return false }
        return currentPeople < peopleCapacity
    }

    func allowsActivatingSpace(currentActiveSpaces: Int) -> Bool {
        guard let activeSpaceCapacity, activeSpaceCapacity > 0 else { return false }
        return currentActiveSpaces < activeSpaceCapacity
    }

    func allowsModule(_ moduleID: String) -> Bool {
        enabledModuleIDs.contains(moduleID)
    }

    func allowsStorageIncrease(currentBytes: Int64, additionalBytes: Int64) -> Bool {
        guard let mediaStorageCapacityBytes,
              currentBytes >= 0,
              additionalBytes >= 0 else {
            return false
        }
        return currentBytes <= mediaStorageCapacityBytes - additionalBytes
    }
}

struct OrganizationUsage: Hashable {
    let peopleCount: Int
    let activeSpaceCount: Int
    let mediaStorageBytes: Int64
}

struct OrganizationOwnedSpace: Identifiable, Hashable {
    let id: String
    let name: String
    let emoji: String
    let memberCount: Int
    let memberIDs: [String]
    let isArchived: Bool
}

struct OrganizationInvite: Identifiable, Hashable {
    let id: String
    let organizationID: String
    let organizationName: String
    let role: OrganizationRole
}
